kubernetes/pods.go

```diff
@@ -0,0 +1,100 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apisv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+const resyncInterval = 1 * time.Minute
+
+// PodInformer is a event handler for Pod events registered to, that builds a local list of valid and relevant pods
+// and sends an event to the endpoint channel, triggering a resync of the targets.
+func (a *Adapter) PodInformer(ctx context.Context, endpointChan chan<- []string) error {
+	if a.cniPodNamespace+a.cniPodLabelSelector == "" {
+		return fmt.Errorf("error scheduling PodInformer: namespace, service, and service port are required")
+	}
+	podEndpoints := sync.Map{}
+
+	log.Infof("Watching for Pods with labelselector %s in namespace %s", a.cniPodNamespace, a.cniPodLabelSelector)
+	factory := informers.NewSharedInformerFactoryWithOptions(a.clientset, resyncInterval, informers.WithNamespace(a.cniPodNamespace),
+		informers.WithTweakListOptions(func(options *apisv1.ListOptions) { options.LabelSelector = a.cniPodLabelSelector }))
+
+	informer := factory.Core().V1().Pods().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newResource interface{}) {
+			pod, ok := newResource.(*corev1.Pod)
+			if !ok {
+				log.Printf("cannot cast object to corev1.Pod %v", newResource)
+				return
+			}
+			switch {
+
+			case pod.DeletionTimestamp != nil:
+				name, exists := podEndpoints.LoadAndDelete(pod.Name)
+				if !exists {
+					return
+				}
+				log.Infof("Deleted pod: %s IP: %s", pod.Name, name)
+				queueEndpoints(&podEndpoints, endpointChan)
+
+			case pod.Status.ContainerStatuses != nil && len(pod.Status.ContainerStatuses) > 0 &&
+				pod.Status.ContainerStatuses[0].State.Running != nil && pod.Status.PodIP != "":
```

Comment:
Would it be better to poll a service endpoints API directly (assuming there is a service that selects proper pods) instead of pods and let k8s to update endpoints when pods get ready?

Reply:
That was my first iteration, but as @szuecs reminded is that a service is normally not deployed with Skipper. Hence I refactored to use pod status. 
BTW although this looks weird, is is the kubectl way to describe the pod status especially to catch the "terminating" state

Reply:
> but as @szuecs reminded is that a service is normally not deployed with Skipper

Having a service could be set as a requirement for CNI mode though

Reply:
There is a principle downside of using service endpoints that they update _after_ the pod readiness probe is true, which stacks up the registration delay on top of the readiness delay. Say 50s initial readiness delay plus 30s registration delay and we are back to the original problem.
The current way registers the target at the same time as the pod is scheduled. The readiness probe is equal to the target health check, hence the pod is online receiving traffic the same time it is marked ready in the cluster.
This way we are saving a lot of time that is crucial when nodes terminate for spot interruptions